### PR TITLE
feat(decide): add a new set of decide apis

### PR DESCRIPTION
**Pull request by jaeopt**
Add a new set of decide apis
- add OptimizelyUserContext (setAttribute/decide/decideAll)
- add OptimizelyDecision
- add "createUserContext" to OptimizelyClient

### CODE REVIEW

File: Sources/Optimizely+Decide/OptimizelyUserContext.swift [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+* See the License for the specific language governing permissions and      *
+* limitations under the License.                                           *
+***************************************************************************/
+    
+import Foundation
+
+/// An object for user contexts that the SDK will use to make decisions for.
+public class OptimizelyUserContext {
+    weak var optimizely: OptimizelyClient?
+    var userId: String
```
Comment (yasirfolio3) [inline Sources/Optimizely+Decide/OptimizelyUserContext.swift:22]: userId and attributes should be public.

Comment (jaeopt) [reply Sources/Optimizely+Decide/OptimizelyUserContext.swift:22]: @yasirfolio3 Sounds like you need to access them directly for FSC? public getter access is enough? Any case you need to directly set?

Comment (yasirfolio3) [reply Sources/Optimizely+Decide/OptimizelyUserContext.swift:22]: yes, no need for setter access, just require public access for FSC to verify the values after each `decide` call 👍

Comment (jaeopt) [reply Sources/Optimizely+Decide/OptimizelyUserContext.swift]: @yasirfolio3 getters added for userId + attributes

**Review by thomaszurkan-optimizely — COMMENTED** [review]
Some of your format changes are exclusive to Swift 5.  Have we decided to no longer support Swift 4 syntactically?  There are a bunch of stuff we can refactor if that is the case.

**Comment by jaeopt** [comment]
> Some of your format changes are exclusive to Swift 5. Have we decided to no longer support Swift 4 syntactically? There are a bunch of stuff we can refactor if that is the case.

We specified the requirement in README "Requirements Swift 5+" for a while.

**Review by thomaszurkan-optimizely — CHANGES_REQUESTED** [review]
Can you fix the show coupon that is no longer working?

**Review by thomaszurkan-optimizely — APPROVED** [review]
I'm changing it to approved with the nit that we get show_coupon working again.  Thanks!

File: Sources/Implementation/DecisionInfo.swift [hunk 2]
```diff
@@ -0,0 +1,150 @@
+//
+/****************************************************************************
+* Copyright 2020, Optimizely, Inc. and contributors                        *
```
Comment (thomaszurkan-optimizely) [inline Sources/Implementation/DecisionInfo.swift:3]: 2021?

File: Sources/Implementation/DecisionReasons.swift [hunk 3]
```diff
@@ -0,0 +1,55 @@
+/****************************************************************************
+* Copyright 2020, Optimizely, Inc. and contributors                        *
```
Comment (thomaszurkan-optimizely) [inline Sources/Implementation/DecisionReasons.swift:2]: ?

File: Sources/Optimizely+Decide/OptimizelyClient+Decide.swift [hunk 4]
```diff
@@ -0,0 +1,204 @@
+/****************************************************************************
+ * Copyright 2020, Optimizely, Inc. and contributors                        *
```
Comment (thomaszurkan-optimizely) [inline Sources/Optimizely+Decide/OptimizelyClient+Decide.swift:2]: 2021?

File: Tests/OptimizelyTests-Common/OptimizelyUserContextTests.swift [hunk 5]
```diff
@@ -0,0 +1,208 @@
+/****************************************************************************
+* Copyright 2020, Optimizely, Inc. and contributors                        *
```
Comment (thomaszurkan-optimizely) [inline Tests/OptimizelyTests-Common/OptimizelyUserContextTests.swift:2]: 2021?

**Review by thomaszurkan-optimizely — APPROVED** [review]
LGTM... nit.. headers are 2020

File: Tests/OptimizelyTests-Common/OptimizelyUserContextTests.swift (hunk 5 above)
Comment (jaeopt) [reply Tests/OptimizelyTests-Common/OptimizelyUserContextTests.swift]: Fixed all copyright years

**Review by thomaszurkan-optimizely — APPROVED** [review]
LGTM

**Pull request merged**